### DISCUSSION TITLE
How to detect Transaction?

### DISCUSSION THREAD

**Discussion by aldycool** [simolus3/drift#1341] [Q&A] [answered]
Hi,

I'm currently building a library that works on top of Moor. One of my function is to execute several Moor's methods atomically, such as `select(table).getSingle()` and continued with `into(table).insert()`, naturally a Transaction will be used here to guarantee the atomic result. I need to check whether my function calls are wrapped inside a Transaction or not, and will raise an error if it isn't. So, how to detect if a certain database subclass (subclassed from `GeneratedDatabase` like the standard example) is already running inside Transaction? I cannot seem to find this in the official documentation, or in the issues. Thanks!

**EDIT**: At the moment, I found a (*hackish*) way of doing this like the following:
```
  bool inTransaction() {
    return Zone.current[#DatabaseConnectionUser]
        .toString()
        .contains("Transaction");
  }
```
In the mean time I'll stick with this way, unless there are more "official" way of doing this. Thanks!

**Comment by simolus3** [comment, marked as answer]
I think your approach is reasonable. Whether something is running in a transaction or not is not currently exposed by moor's public api, but I'll take a look at adding that information.

**Reply by simolus3** [reply]
I've made the information a bit more public in 159656bc74387cbbdb4caad51afe4bccb8c5199c. You can use `resolvedEngine is Transaction` to detect whether the current context is in a transaction or not.
I've currently flagged it as `@internal` since I'm not entirely convinced if this is an appropriate way to expose that information. But you can ignore the lint and just use this, I wont change it before a breaking release.

**Reply by aldycool** [reply]
Alright, I think by checking the `resolvedEngine` feels more "concrete". Thank you for your consideration.
